### ISSUE TITLE
Quickly convert DBObject instances to ordinary objects

### ISSUE THREAD

**Issue by LeavesSky** [oracle/node-oracledb#1184]
Environments :
1. Node.js                v12.6.0
2. OracleDB             v4.1
3. Oracle                 11g R2 64 bit
4. about code:        Query Oracle Spatial

Because the stock data is not standardized and needs to be traversed once and processed accordingly (offset and structure are `GeoSJON` ), I return the field of type `SDO_Geometry` directly to the result set. It is found that the conversion efficiency of multiple data is particularly low
The scheme I tried is as follows (my offset and structure are synchronous code, which is highly efficient and the impact can be ignored):

1. Check your JS code and find the `toJSON` method of `DBObject`. The serialization is slow;
2. `JSON.parse(JSON.stringify( geometry ))`;
3. Upper `getvalues()` of `DBObject` instance;
4. Try to use the new multithreading of nodejs12;
5. Divide the result set into blocks, and then use async to process;

After the above attempts, I found that: the result set is mounted on the 'connection', because as long as the 'connection' is released, the result set cannot continue to be parsed; connection is a single instance, which cannot improve the intensive computation of serialization;

Today, I added `JSON. Parse (JSON. Stringify (rows))` to the `fetchrowscb` function in `connection. JS`; the efficiency has been increased by 10 times;
```javascript
  const fetchRowsCb = function (err, rows) {

    if (err) {
      cb(err);
      return;
    }

    if (rows) {
      rowsFetched = rowsFetched.concat(JSON.parse(JSON.stringify(rows)));
    }

    if (rowsFetched.length == maxRows || rows.length < fetchArraySize) {
      cb(null, rowsFetched);
      return;
    }

    resultSet.getRows(fetchArraySize, fetchRowsCb);
  };
```
I hope to get a result set of ordinary objects;

Thank you

**Comment by anthony-tuininga** [comment]
Thanks, @LeavesSky. We were already aware that the transformation from database objects to plain JavaScript objects is not particularly fast. It is on our list of things to look at. There are instances where you want the database backed object and instances where you want just a plain JavaScript object. So we'll need some way for you to specify which one you want. We'll give it some thought and get back to you!

**Comment by LeavesSky** [comment]
Thanks,&nbsp;I sincerely look forward to your reply



------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Anthony Tuininga"<notifications@github.com&gt;; 
发送时间: 2019年12月6日(星期五) 晚上11:57
收件人: "oracle/node-oracledb"<node-oracledb@noreply.github.com&gt;; 
抄送: "╰つ℡ 吃豆人╰╮"<1559935142@qq.com&gt;; "Mention"<mention@noreply.github.com&gt;; 
主题: Re: [oracle/node-oracledb] Quickly convert DBObject instances to ordinary objects (#1184)




Thanks, @LeavesSky. We were already aware that the transformation from database objects to plain JavaScript objects is not particularly fast. It is on our list of things to look at. There are instances where you want the database backed object and instances where you want just a plain JavaScript object. So we'll need some way for you to specify which one you want. We'll give it some thought and get back to you!

**Comment by anthony-tuininga** [comment]
I've been looking at this and trying to follow what you are doing to improve performance. Just fetching the rows vs calling `JSON.parse(JSON.stringify(rows))` is obviously faster. In other words, the fetch itself isn't slow. It depends on what you are doing afterwards that matters. Can you clarify?

**Comment by LeavesSky** [comment]
Scenario: get data of `sdo_geometry` type from database, and then extend and convert it to geojson structure data;
When the amount of data reaches more than 10000, it is obviously slow. Personal Troubleshooting: it takes a long time to get data from `dbobject`;
Therefore, there are many attempts above;

**Comment by cjbj** [comment]
@LeavesSky what are you doing with the data in general?  What is "obviously slow"?  Why would conversion in node-oracledb would be any faster than you doing it in the app - the oracle object still has to be traversed?  Can you give a runnable example that shows how you were handling connections, getting data, and using that data?

**Comment by LeavesSky** [comment]
Environments 

1. Node.js v12.16.1
2. OracleDB v4.2
3. Oracle 11g R2 64 bit
4. about code: Query Oracle Spatial
---
The following are the code blocks involved in the problem（Convert dbobject to normal JavaScript object inside the module）:
```
const geoArr = rows.reduce((pv, cv) => {
  const shapeInfo = cv[shapeIndex];
  const elem_info = [];
  if (shapeInfo.SDO_ELEM_INFO && shapeInfo.SDO_ELEM_INFO.length > 3) {
    elem_info.push(...multi_elem_Info(shapeInfo.SDO_ELEM_INFO));
  }
  if (shapeInfo.SDO_GTYPE < 1000) shapeInfo.SDO_GTYPE += 2000;
  else if (shapeInfo.SDO_GTYPE < 2000) shapeInfo.SDO_GTYPE += 1000;
  let type = typeMetadataStrs[typeMetadataNums.indexOf(shapeInfo.SDO_GTYPE)];
  if (GType.includes(shapeInfo.SDO_GTYPE)) {
    const data = properties.reduce((obj, cn) => {
      const vi = metaData.findIndex(clo => clo.name === cn);
      obj[cn.toLocaleLowerCase()] = cv[vi];
      return obj;
    }, {});
    if (typeof transform === 'object' && transform.state && typeMetadataStrs.includes(transform.new) && typeMetadataStrs.includes(transform.old)) {
      if (transform.new === 'Point') {
        type = 'Point';
        if (Array.isArray(shapeInfo.SDO_ORDINATES)) {
          if (shapeInfo.SDO_ORDINATES.length % 2) shapeInfo.SDO_ORDINATES.pop();
          const dividend = shapeInfo.SDO_ORDINATES.length / 2;
          const { odd, even } = shapeInfo.SDO_ORDINATES.reduce((numObj, cc, ii) => {
            if (ii % 2) {
              numObj.odd += cc;
            } else {
              numObj.even += cc;
            }
            return numObj;
          }, { odd: 0, even: 0 });
          data[type] = coordinatesTransform([ even / dividend, odd / dividend ], type, {
            offsetX,
            offsetY,
          });
        } else {
          data[type] = [];
        }
      }
    } else {
      if (elem_info && elem_info.length > 1) {
        data[type] = coordinatesTransform(shapeInfo.SDO_ORDINATES, type, {
          offsetX,
          offsetY,
        }, elem_info);
      } else {
        if (type === 'Point' && shapeInfo.SDO_POINT && shapeInfo.SDO_POINT.X && shapeInfo.SDO_POINT.Y && verifylong(shapeInfo.SDO_POINT.X) && verifylat(shapeInfo.SDO_POINT.Y)) shapeInfo.SDO_ORDINATES = [ shapeInfo.SDO_POINT.X, shapeInfo.SDO_POINT.Y ];
        if (Array.isArray(shapeInfo.SDO_ORDINATES)) {
          data[type] = coordinatesTransform(shapeInfo.SDO_ORDINATES, type, {
            offsetX,
            offsetY,
          });
        } else {
          data[type] = [];
        }
      }
      if (data.sbtyys && initStyle) data.style = initStyle[data.sbtyys].WBKZBY5;
    }
    coordinates.push(data[type]);
    pv.push(data);
  }
  return pv;
}, []);
GeoJSON.parse(geoArr, geoOptions);
```

Upgrade to version node-oracledb@4.2, which can be used like ordinary JavaScript objects, but still slow when the data volume is large

**Comment by cjbj** [comment]
Thank you

**Comment by cjbj** [comment]
@LeavesSky see the `dbObjectAsPojo` feature that I just pushed to the master branch.  Let us know if this works for you. The (development) doc is [here](https://github.com/oracle/node-oracledb/blob/master/doc/api.md#propdbobjpojo). See [here](https://oracle.github.io/node-oracledb/INSTALL.html#github) for how to build from GitHub.

**Comment by cjbj** [comment]
This was released in node-oracledb 5.1.
